File: sdks/typescript/src/geospatial.ts
/**
 * Synap TypeScript SDK - Geospatial Manager
 * 
 * Redis-compatible Geospatial operations for location-based queries
 */

import { SynapClient } from './client';
import type { CommandOptions } from './types';

/**
 * Distance unit types
 */
export type DistanceUnit = 'm' | 'km' | 'mi' | 'ft';

/**
 * Geospatial statistics
 */
export interface GeospatialStats {
  total_keys: number;
  total_locations: number;
  geoadd_count: number;
  geodist_count: number;
  georadius_count: number;
  geopos_count: number;
  geohash_count: number;
}

/**
 * Location coordinate
 */
export interface Location {
  lat: number;
  lon: number;
  member: string;
}

/**
 * Geospatial coordinate
 */
export interface Coordinate {
  lat: number;
  lon: number;
}

/**
 * Georadius result item
 */
export interface GeoradiusResult {
  member: string;
  distance?: number;
  coord?: Coordinate;
}

interface GeoaddResponse {
  key: string;
  added: number;
}

interface GeodistResponse {
  key: string;
  distance: number | null;
  unit: string;
}

interface GeoradiusResponse {
  key: string;
  results: GeoradiusResult[];
}

interface GeoposResponse {
  key: string;
  coordinates: (Coordinate | null)[];
}

interface GeohashResponse {
  key: string;
  geohashes: (string | null)[];
}

/**
 * Geospatial operations manager
 */
export class GeospatialManager {
  constructor(private readonly client: SynapClient) {}

  private buildPayload(
    options: CommandOptions | undefined,
    extra: Record<string, unknown>
  ): Record<string, unknown> {
    const payload: Record<string, unknown> = { ...extra };

    if (options?.clientId) {
      payload.client_id = options.clientId;
    }

    return payload;
  }

  /**
   * Add geospatial locations (GEOADD)
   * @param key Geospatial key
   * @param locations Array of locations (lat, lon, member)
   * @param options Optional options (nx, xx, ch)
   * @returns Number of elements added
   */
  async geoadd(
    key: string,
    locations: Location[],
    options?: CommandOptions & { nx?: boolean; xx?: boolean; ch?: boolean }
  ): Promise<number> {
    // Validate coordinates
    for (const loc of locations) {
      if (loc.lat < -90 || loc.lat > 90) {
        throw new TypeError(`Latitude must be between -90 and 90, got: ${loc.lat}`);
      }
      if (loc.lon < -180 || loc.lon > 180) {
        throw new TypeError(`Longitude must be between -180 and 180, got: ${loc.lon}`);
      }
    }

    const response = await this.client.sendCommand<GeoaddResponse>(
      'geospatial.geoadd',
      this.buildPayload(options, {
        key,
        locations,
        nx: options?.nx ?? false,
        xx: options?.xx ?? false,
        ch: options?.ch ?? false,
      })
    );

    return response.added;
  }

  /**
   * Calculate distance between two members (GEODIST)
   * @param key Geospatial key
   * @param member1 First member
   * @param member2 Second member
   * @param unit Distance unit (default: 'm')
   * @param options Optional command options
   * @returns Distance in specified unit, or null if either member doesn't exist
   */
  async geodist(
    key: string,
    member1: string,
    member2: string,
    unit: DistanceUnit = 'm',
    options?: CommandOptions
  ): Promise<number | null> {
    const response = await this.client.sendCommand<GeodistResponse>(
      'geospatial.geodist',
      this.buildPayload(options, {
        key,
        member1,
        member2,
        unit,
      })
    );

    return response.distance;
  }

  /**
   * Query members within radius (GEORADIUS)
   * @param key Geospatial key
   * @param centerLat Center latitude
   * @param centerLon Center longitude
   * @param radius Radius
   * @param unit Distance unit (default: 'm')
   * @param options Query options (withDist, withCoord, count, sort)
   * @param commandOptions Optional command options
   * @returns Array of matching members with optional distance and coordinates
   */
  async georadius(
    key: string,
    centerLat: number,
    centerLon: number,
    radius: number,
    unit: DistanceUnit = 'm',
    options?: {
      withDist?: boolean;
      withCoord?: boolean;
      count?: number;
      sort?: 'ASC' | 'DESC';
    },
    commandOptions?: CommandOptions
  ): Promise<GeoradiusResult[]> {
    if (centerLat < -90 || centerLat > 90) {
      throw new TypeError(`Latitude must be between -90 and 90, got: ${centerLat}`);
    }
    if (centerLon < -180 || centerLon > 180) {
      throw new TypeError(`Longitude must be between -180 and 180, got: ${centerLon}`);
    }

    const response = await this.client.sendCommand<GeoradiusResponse>(
      'geospatial.georadius',
      this.buildPayload(commandOptions, {
        key,
        center_lat: centerLat,
        center_lon: centerLon,
        radius,
        unit,
        with_dist: options?.withDist ?? false,
        with_coord: options?.withCoord ?? false,
        count: options?.count,
        sort: options?.sort,
      })
    );

    return response.results;
  }

  /**
   * Query members within radius of given member (GEORADIUSBYMEMBER)
   * @param key Geospatial key
   * @param member Center member
   * @param radius Radius
   * @param unit Distance unit (default: 'm')
   * @param options Query options (withDist, withCoord, count, sort)
   * @param commandOptions Optional command options
   * @returns Array of matching members with optional distance and coordinates
   */
  async georadiusbymember(
    key: string,
    member: string,
    radius: number,
    unit: DistanceUnit = 'm',
    options?: {
      withDist?: boolean;
      withCoord?: boolean;
      count?: number;
      sort?: 'ASC' | 'DESC';
    },
    commandOptions?: CommandOptions
  ): Promise<GeoradiusResult[]> {
    const response = await this.client.sendCommand<GeoradiusResponse>(
      'geospatial.georadiusbymember',
      this.buildPayload(commandOptions, {
        key,
        member,
        radius,
        unit,
        with_dist: options?.withDist ?? false,
        with_coord: options?.withCoord ?? false,
        count: options?.count,
        sort: options?.sort,
      })
    );

    return response.results;
  }

  /**
   * Get coordinates of members (GEOPOS)
   * @param key Geospatial key
   * @param members Array of member names
   * @param options Optional command options
   * @returns Array of coordinates (null if member doesn't exist)
   */
  async geopos(
    key: string,
    members: string[],
    options?: CommandOptions
  ): Promise<(Coordinate | null)[]> {
    const response = await this.client.sendCommand<GeoposResponse>(
      'geospatial.geopos',
      this.buildPayload(options, {
        key,
        members,
      })
    );

    return response.coordinates;
  }

  /**
   * Get geohash strings for members (GEOHASH)
   * @param key Geospatial key
   * @param members Array of member names
   * @param options Optional command options
   * @returns Array of geohash strings (null if member doesn't exist)
   */
  async geohash(
    key: string,
    members: string[],
    options?: CommandOptions
  ): Promise<(string | null)[]> {
    const response = await this.client.sendCommand<GeohashResponse>(
      'geospatial.geohash',
      this.buildPayload(options, {
        key,
        members,
      })
    );

    return response.geohashes;
  }

  /**
   * Advanced geospatial search (GEOSEARCH)
   * @param key Geospatial key
   * @param options Search options (fromMember/fromLonLat, byRadius/byBox, withDist, withCoord, count, sort)
   * @param commandOptions Optional command options
   * @returns Array of matching members with optional distance and coordinates
   */
  async geosearch(
    key: string,
    options: {
      fromMember?: string;
      fromLonLat?: [number, number]; // [lon, lat]
      byRadius?: [number, DistanceUnit];
      byBox?: [number, number, DistanceUnit]; // [width, height, unit]
      withDist?: boolean;
      withCoord?: boolean;
      withHash?: boolean;
      count?: number;
      sort?: 'ASC' | 'DESC';
    },
    commandOptions?: CommandOptions
  ): Promise<GeoradiusResult[]> {
    if (!options.fromMember && !options.fromLonLat) {
      throw new TypeError("Either 'fromMember' or 'fromLonLat' must be provided");
    }
    if (!options.byRadius && !options.byBox) {
      throw new TypeError("Either 'byRadius' or 'byBox' must be provided");
    }

    const payload: Record<string, unknown> = {
      key,
      with_dist: options.withDist ?? false,
      with_coord: options.withCoord ?? false,
      with_hash: options.withHash ?? false,
    };

    if (options.fromMember) {
      payload.from_member = options.fromMember;
    }
    if (options.fromLonLat) {
      payload.from_lonlat = options.fromLonLat;
    }
    if (options.byRadius) {
      payload.by_radius = [options.byRadius[0], options.byRadius[1]];
    }
    if (options.byBox) {
      payload.by_box = [options.byBox[0], options.byBox[1], options.byBox[2]];
    }
    if (options.count !== undefined) {
      payload.count = options.count;
    }
    if (options.sort) {
      payload.sort = options.sort;
    }

    const response = await this.client.sendCommand<GeoradiusResponse>(
      'geospatial.geosearch',
      this.buildPayload(commandOptions, payload)
    );

    return response.results;
  }

  /**
   * Retrieve geospatial statistics
   * @param options Optional command options
   * @returns Geospatial statistics
   */
  async stats(options?: CommandOptions): Promise<GeospatialStats> {
    const response = await this.client.sendCommand<GeospatialStats>(
      'geospatial.stats',
      this.buildPayload(options, {})
    );

    return response;
  }
}

File: sdks/rust/src/geospatial.rs
//! Geospatial operations (GEOADD/GEODIST/GEORADIUS/GEOPOS/GEOHASH)

use crate::client::SynapClient;
use crate::error::Result;
use serde::{Deserialize, Serialize};
use serde_json::json;

/// Distance unit
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DistanceUnit {
    Meters,
    Kilometers,
    Miles,
    Feet,
}

impl DistanceUnit {
    fn as_str(&self) -> &'static str {
        match self {
            DistanceUnit::Meters => "m",
            DistanceUnit::Kilometers => "km",
            DistanceUnit::Miles => "mi",
            DistanceUnit::Feet => "ft",
        }
    }
}

/// Location coordinate
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Location {
    pub lat: f64,
    pub lon: f64,
    pub member: String,
}

/// Coordinate
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Coordinate {
    pub lat: f64,
    pub lon: f64,
}

/// Georadius result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GeoradiusResult {
    pub member: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub distance: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub coord: Option<Coordinate>,
}

/// Geospatial statistics
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct GeospatialStats {
    pub total_keys: usize,
    pub total_locations: usize,
    pub geoadd_count: usize,
    pub geodist_count: usize,
    pub georadius_count: usize,
    pub geopos_count: usize,
    pub geohash_count: usize,
}

#[derive(Clone)]
pub struct GeospatialManager {
    client: SynapClient,
}

impl GeospatialManager {
    pub(crate) fn new(client: SynapClient) -> Self {
        Self { client }
    }

    /// Add geospatial locations (GEOADD)
    ///
    /// # Arguments
    ///
    /// * `key` - Geospatial key
    /// * `locations` - Array of locations (lat, lon, member)
    /// * `nx` - Only add new elements (don't update existing)
    /// * `xx` - Only update existing elements (don't add new)
    /// * `ch` - Return count of changed elements
    ///
    /// # Returns
    ///
    /// Number of elements added
    pub async fn geoadd(
        &self,
        key: &str,
        locations: Vec<Location>,
        nx: bool,
        xx: bool,
        ch: bool,
    ) -> Result<usize> {
        // Validate coordinates
        for loc in &locations {
            if !(-90.0..=90.0).contains(&loc.lat) {
                return Err(crate::error::SynapError::ServerError(format!(
                    "Latitude must be between -90 and 90, got: {}",
                    loc.lat
                )));
            }
            if !(-180.0..=180.0).contains(&loc.lon) {
                return Err(crate::error::SynapError::ServerError(format!(
                    "Longitude must be between -180 and 180, got: {}",
                    loc.lon
                )));
            }
        }

        let payload = json!({
            "key": key,
            "locations": locations,
            "nx": nx,
            "xx": xx,
            "ch": ch,
        });

        let response = self
            .client
            .send_command("geospatial.geoadd", payload)
            .await?;
        Ok(response["added"].as_u64().unwrap_or(0) as usize)
    }

    /// Calculate distance between two members (GEODIST)
    ///
    /// # Arguments
    ///
    /// * `key` - Geospatial key
    /// * `member1` - First member
    /// * `member2` - Second member
    /// * `unit` - Distance unit
    ///
    /// # Returns
    ///
    /// Distance in specified unit, or None if either member doesn't exist
    pub async fn geodist(
        &self,
        key: &str,
        member1: &str,
        member2: &str,
        unit: DistanceUnit,
    ) -> Result<Option<f64>> {
        let payload = json!({
            "key": key,
            "member1": member1,
            "member2": member2,
            "unit": unit.as_str(),
        });

        let response = self
            .client
            .send_command("geospatial.geodist", payload)
            .await?;
        let distance = response["distance"].as_f64();
        Ok(distance)
    }

    /// Query members within radius (GEORADIUS)
    ///
    /// # Arguments
    ///
    /// * `key` - Geospatial key
    /// * `center_lat` - Center latitude
    /// * `center_lon` - Center longitude
    /// * `radius` - Radius
    /// * `unit` - Distance unit
    /// * `with_dist` - Include distance in results
    /// * `with_coord` - Include coordinates in results
    /// * `count` - Maximum number of results
    /// * `sort` - Sort order ("ASC" or "DESC")
    ///
    /// # Returns
    ///
    /// Vector of matching members with optional distance and coordinates
    #[allow(clippy::too_many_arguments)]
    pub async fn georadius(
        &self,
        key: &str,
        center_lat: f64,
        center_lon: f64,
        radius: f64,
        unit: DistanceUnit,
        with_dist: bool,
        with_coord: bool,
        count: Option<usize>,
        sort: Option<&str>,
    ) -> Result<Vec<GeoradiusResult>> {
        if !(-90.0..=90.0).contains(&center_lat) {
            return Err(crate::error::SynapError::ServerError(format!(
                "Latitude must be between -90 and 90, got: {}",
                center_lat
            )));
        }
        if !(-180.0..=180.0).contains(&center_lon) {
            return Err(crate::error::SynapError::ServerError(format!(
                "Longitude must be between -180 and 180, got: {}",
                center_lon
            )));
        }

        let mut payload = json!({
            "key": key,
            "center_lat": center_lat,
            "center_lon": center_lon,
            "radius": radius,
            "unit": unit.as_str(),
            "with_dist": with_dist,
            "with_coord": with_coord,
        });

        if let Some(c) = count {
            payload["count"] = json!(c);
        }
        if let Some(s) = sort {
            payload["sort"] = json!(s);
        }

        let response = self
            .client
            .send_command("geospatial.georadius", payload)
            .await?;
        let results: Vec<GeoradiusResult> =
            serde_json::from_value(response["results"].clone()).unwrap_or_default();
        Ok(results)
    }

    /// Query members within radius of given member (GEORADIUSBYMEMBER)
    ///
    /// # Arguments
    ///
    /// * `key` - Geospatial key
    /// * `member` - Center member
    /// * `radius` - Radius
    /// * `unit` - Distance unit
    /// * `with_dist` - Include distance in results
    /// * `with_coord` - Include coordinates in results
    /// * `count` - Maximum number of results
    /// * `sort` - Sort order ("ASC" or "DESC")
    ///
    /// # Returns
    ///
    /// Vector of matching members with optional distance and coordinates
    #[allow(clippy::too_many_arguments)]
    pub async fn georadiusbymember(
        &self,
        key: &str,
        member: &str,
        radius: f64,
        unit: DistanceUnit,
        with_dist: bool,
        with_coord: bool,
        count: Option<usize>,
        sort: Option<&str>,
    ) -> Result<Vec<GeoradiusResult>> {
        let mut payload = json!({
            "key": key,
            "member": member,
            "radius": radius,
            "unit": unit.as_str(),
            "with_dist": with_dist,
            "with_coord": with_coord,
        });

        if let Some(c) = count {
            payload["count"] = json!(c);
        }
        if let Some(s) = sort {
            payload["sort"] = json!(s);
        }

        let response = self
            .client
            .send_command("geospatial.georadiusbymember", payload)
            .await?;
        let results: Vec<GeoradiusResult> =
            serde_json::from_value(response["results"].clone()).unwrap_or_default();
        Ok(results)
    }

    /// Get coordinates of members (GEOPOS)
    ///
    /// # Arguments
    ///
    /// * `key` - Geospatial key
    /// * `members` - Array of member names
    ///
    /// # Returns
    ///
    /// Vector of coordinates (None if member doesn't exist)
    pub async fn geopos(&self, key: &str, members: &[String]) -> Result<Vec<Option<Coordinate>>> {
        let payload = json!({
            "key": key,
            "members": members,
        });

        let response = self
            .client
            .send_command("geospatial.geopos", payload)
            .await?;
        let coords: Vec<Option<Coordinate>> =
            serde_json::from_value(response["coordinates"].clone()).unwrap_or_default();
        Ok(coords)
    }

    /// Advanced geospatial search (GEOSEARCH)
    ///
    /// # Arguments
    ///
    /// * `key` - Geospatial key
    /// * `from_member` - Center member (mutually exclusive with from_lonlat)
    /// * `from_lonlat` - Center coordinates as (lon, lat) tuple (mutually exclusive with from_member)
    /// * `by_radius` - Search by radius as (radius, unit) tuple
    /// * `by_box` - Search by bounding box as (width, height, unit) tuple
    /// * `with_dist` - Include distance in results
    /// * `with_coord` - Include coordinates in results
    /// * `with_hash` - Include geohash in results (not yet implemented)
    /// * `count` - Maximum number of results
    /// * `sort` - Sort order ("ASC" or "DESC")
    ///
    /// # Returns
    ///
    /// Vector of matching members with optional distance and coordinates
    #[allow(clippy::too_many_arguments)]
    pub async fn geosearch(
        &self,
        key: &str,
        from_member: Option<&str>,
        from_lonlat: Option<(f64, f64)>,
        by_radius: Option<(f64, DistanceUnit)>,
        by_box: Option<(f64, f64, DistanceUnit)>,
        with_dist: bool,
        with_coord: bool,
        with_hash: bool,
        count: Option<usize>,
        sort: Option<&str>,
    ) -> Result<Vec<GeoradiusResult>> {
        if from_member.is_none() && from_lonlat.is_none() {
            return Err(crate::error::SynapError::ServerError(
                "Either 'from_member' or 'from_lonlat' must be provided".to_string(),
            ));
        }
        if by_radius.is_none() && by_box.is_none() {
            return Err(crate::error::SynapError::ServerError(
                "Either 'by_radius' or 'by_box' must be provided".to_string(),
            ));
        }

        let mut payload = json!({
            "key": key,
            "with_dist": with_dist,
            "with_coord": with_coord,
            "with_hash": with_hash,
        });

        if let Some(member) = from_member {
            payload["from_member"] = json!(member);
        }
        if let Some((lon, lat)) = from_lonlat {
            payload["from_lonlat"] = json!([lon, lat]);
        }
        if let Some((radius, unit)) = by_radius {
            payload["by_radius"] = json!([radius, unit.as_str()]);
        }
        if let Some((width, height, unit)) = by_box {
            payload["by_box"] = json!([width, height, unit.as_str()]);
        }
        if let Some(c) = count {
            payload["count"] = json!(c);
        }
        if let Some(s) = sort {
            payload["sort"] = json!(s);
        }

        let response = self
            .client
            .send_command("geospatial.geosearch", payload)
            .await?;
        let results: Vec<GeoradiusResult> =
            serde_json::from_value(response["results"].clone()).unwrap_or_default();
        Ok(results)
    }

    /// Get geohash strings for members (GEOHASH)
    ///
    /// # Arguments
    ///
    /// * `key` - Geospatial key
    /// * `members` - Array of member names
    ///
    /// # Returns
    ///
    /// Vector of geohash strings (None if member doesn't exist)
    pub async fn geohash(&self, key: &str, members: &[String]) -> Result<Vec<Option<String>>> {
        let payload = json!({
            "key": key,
            "members": members,
        });

        let response = self
            .client
            .send_command("geospatial.geohash", payload)
            .await?;
        let geohashes: Vec<Option<String>> =
            serde_json::from_value(response["geohashes"].clone()).unwrap_or_default();
        Ok(geohashes)
    }

    /// Retrieve geospatial statistics
    pub async fn stats(&self) -> Result<GeospatialStats> {
        let payload = json!({});
        let response = self
            .client
            .send_command("geospatial.stats", payload)
            .await?;
        let stats: GeospatialStats = serde_json::from_value(response).unwrap_or_default();
        Ok(stats)
    }
}
